File: C++/Problems/Problems/problem_2365.cpp
#include "problem_2365.h"

#include <unordered_map>

bool problem_2365::test() {
    std::vector tasks { 1, 2, 1, 2, 3, 1 };
    int space = 3;

    int expected = 9;

    auto result = taskSchedulerII(tasks, space);

    std::cout << result << std::endl;

    return result == expected;
}

long long problem_2365::taskSchedulerII(std::vector<int> &tasks, int space) {
    long long result = 0;
    std::unordered_map<int, long long> th;
    for (auto t : tasks)
    {
        if (!th.contains(t))
            th[t] = 0L;
        result = std::max(th[t], result) + 1;
        th[t] = result + space;
    }

    return result;
}

File: Python/Problems/Problems/problem2365.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem


class Problem2365(Problem):
    def test(self) -> bool:
        tasks = [1, 2, 1, 2, 3, 1]
        space = 3

        expected = 9

        result = self.taskSchedulerII(tasks, space)

        print(result)

        return result == expected

    def taskSchedulerII(self, tasks: List[int], space: int) -> int:
        result = 0
        th = dict()
        for t in tasks:
            if t not in th:
                th[t] = 0
            result = max(th[t], result) + 1
            th[t] = result + space

        return result